Add initial support for matroska / webm

darksv opened the pull request:
Hi! This PR adds an initial support for matroska and webm containers that is intended to close #63. It is far from complete, but I decided to share what I have so far to get some feedback.

pdeljanov commented on symphonia-format-mkv/src/ebml.rs:
```diff
@@ -0,0 +1,303 @@
+                    let value = match header.data_len {
+                        0 => 0.0,
+                        4 => {
+                            let mut buff = [0u8; 4];
+                            reader.read_buf_exact(&mut buff)?;
+                            f32::from_be_bytes(buff).into()
```
You can do `reader.read_be_f32()?`.

pdeljanov commented on symphonia-format-mkv/src/ebml.rs:
```diff
@@ -0,0 +1,303 @@
+                            f32::from_be_bytes(buff).into()
+                        }
+                        8 => {
+                            let mut buff = [0u8; 8];
+                            reader.read_buf_exact(&mut buff)?;
+                            f64::from_be_bytes(buff)
```
`reader.read_be_f64()?`

pdeljanov commented on symphonia-format-mkv/src/ebml.rs:
```diff
@@ -0,0 +1,303 @@
+                    Value::String(std::str::from_utf8(&s).unwrap().to_string())
+                }
+                Type::Binary => {
+                    let mut v = vec![0u8; header.data_len as usize];
+                    reader.read_buf_exact(&mut v)?;
+                    Value::Binary(v.into_boxed_slice())
```
`reader.read_boxed_slice_exact()?`.

pdeljanov commented on symphonia-format-mkv/src/ebml.rs:
```diff
@@ -0,0 +1,303 @@
+                    Value::Float(value)
+                }
+                Type::Unknown | Type::Date => {
+                    // TODO
+                    let mut v = vec![0u8; header.data_len as usize];
+                    reader.read_buf_exact(&mut v)?;
+                    Value::Binary(v.into_boxed_slice())
```
`reader.read_boxed_slice_exact()?`

pdeljanov commented on symphonia-format-mkv/src/element_ids.rs:
```diff
@@ -0,0 +1,306 @@
+    (0x55D5, Type::Float, ElementType::PrimaryBChromaticityX),
+    (0x55D6, Type::Float, ElementType::PrimaryBChromaticityY),
+    (0x55D7, Type::Float, ElementType::WhitePointChromaticityX),
+    (0x55D8, Type::Float, ElementType::WhitePointChromaticityY),
+    (0x55D9, Type::Float, ElementType::LuminanceMax),
+    (0x55DA, Type::Float, ElementType::LuminanceMin),
+    (0x1C53BB6B, Type::Master, ElementType::Cues),
+    (0xBB, Type::Master, ElementType::CuePoint),
+    (0xB3, Type::Unsigned, ElementType::CueTime),
+    (0xB7, Type::Master, ElementType::CueTrackPositions),
+    (0xF7, Type::Unsigned, ElementType::CueTrack),
+    (0xF1, Type::Unsigned, ElementType::CueClusterPosition),
+    (0xF0, Type::Unsigned, ElementType::CueRelativePosition),
+    (0xB2, Type::Unsigned, ElementType::CueDuration),
+    (0x5378, Type::Unsigned, ElementType::CueBlockNumber),
+    (0x1043A770, Type::Master, ElementType::Chapters),
+    (0x45B9, Type::Master, ElementType::EditionEntry),
+    (0xB6, Type::Master, ElementType::ChapterAtom),
+    (0x73C4, Type::Unsigned, ElementType::ChapterUid),
+    (0x5654, Type::String, ElementType::ChapterStringUid),
+    (0x91, Type::Unsigned, ElementType::ChapterTimeStart),
+    (0x92, Type::Unsigned, ElementType::ChapterTimeEnd),
+    (0x80, Type::Master, ElementType::ChapterDisplay),
+    (0x85, Type::String, ElementType::ChapString),
+    (0x437C, Type::String, ElementType::ChapLanguage),
+    (0x437D, Type::String, ElementType::ChapLanguageIetf),
+    (0x437E, Type::String, ElementType::ChapCountry),
+    (0x1254C367, Type::Master, ElementType::Tags),
+    (0x7373, Type::Master, ElementType::Tag),
+    (0x63C0, Type::Master, ElementType::Targets),
+    (0x68CA, Type::Unsigned, ElementType::TargetTypeValue),
+    (0x63CA, Type::String, ElementType::TargetType),
+    (0x63C5, Type::Unsigned, ElementType::TagTrackUid),
+    (0x67C8, Type::Master, ElementType::SimpleTag),
+    (0x45A3, Type::String, ElementType::TagName),
+    (0x447A, Type::String, ElementType::TagLanguage),
+    (0x4484, Type::Unsigned, ElementType::TagDefault),
+    (0x4487, Type::String, ElementType::TagString),
+    (0x4485, Type::Binary, ElementType::TagBinary),
+];
```
If you need to do a lookup based on the integer value, consider using `lazy_static` to initialize a `HashMap` or `BTreeMap` instead. This'll make the lookup faster. I did a similar thing for the ID3v2 parser [here](https://github.com/pdeljanov/Symphonia/blob/6e4653c6a376a0f70ac987b3f9579c54d6134b53/symphonia-metadata/src/id3v2/frames.rs#L240-L353).

pdeljanov commented on symphonia-format-mkv/src/lib.rs:
```diff
@@ -0,0 +1,220 @@
+
+        let tracks: Vec<_> = segment.tracks.into_vec().into_iter().map(|track| Track {
+            id: track.id as u32,
+            codec_params: CodecParameters {
+                codec: CODEC_TYPE_OPUS,
+                sample_rate: track.audio.map(|it| it.sampling_frequency.round() as u32),
+                time_base: None,
+                n_frames: None,
+                start_ts: 0,
+                sample_format: Some(SampleFormat::S16),
+                bits_per_sample: None,
+                bits_per_coded_sample: None,
+                channels: Some(Channels::SIDE_LEFT | Channels::SIDE_RIGHT),
+                channel_layout: Some(Layout::Stereo),
+                leading_padding: None,
+                trailing_padding: None,
+                max_frames_per_packet: None,
+                packet_data_integrity: false,
+                verification_check: None,
+                extra_data: track.codec_private,
+            },
```
`CodecParameters` has a builder-like interface. You can do something like the following rather than specifying each field manually.

```
let mut params = CodecParameter::new();

params.for_codec(CODEC_TYPE_OPUS)
      .with_sample_rate(track.audio.map(|it| it.sampling_frequency.round() as u32))
      .with_sample_format(SampleFormat::S16);
```

pdeljanov reviewed (changes requested):
I gave this a quick test by enabling `mkv` in `symphonia-play` and I can see it correctly exposing the right number of tracks for the file. This is really fantastic start!

In addition to the issues noted in the review, please also make sure to put yourself as the author in the `Cargo.toml` file for the demuxer, and add yourself to the `CONTRIBUTORS` file in the project root.

darksv commented on symphonia-core/src/meta.rs:
```diff
@@ -260,6 +260,29 @@ impl_from_for_value!(v, &str, Value::String(String::from(v)));
 impl_from_for_value!(v, String, Value::String(v));
 impl_from_for_value!(v, Cow<'_, str>, Value::String(String::from(v)));
 
+impl Value {
+    pub fn to_u64(&self) -> Option<u64> {
+        match self {
+            Value::UnsignedInt(value) => Some(*value),
+            _ => None,
+        }
+    }
+
+    pub fn to_f64(&self) -> Option<f64> {
+        match self {
+            Value::Float(value) => Some(*value),
+            _ => None,
+        }
+    }
+
+    pub fn to_bytes(&self) -> Option<&[u8]> {
+        match self {
+            Value::Binary(value) => Some(value),
+            _ => None,
+        }
+    }
+}
+
```
Yeah, I missed the docs saying that it's for tags. I'll add a new enum 😅

darksv commented:
I tested it using `symphonia-play` and it seems to work on a few files already... 😄

pdeljanov commented on symphonia-format-mkv/src/codecs.rs:
```diff
@@ -0,0 +1,36 @@
+            64 => Some(codecs::CODEC_TYPE_PCM_F32BE),
+            _ => None,
+        }
+        // TODO
```
Could we add support for Vorbis and AAC codecs here too? I believe these codecs correspond to `A_VORBIS`, and `A_AAC/MPEG4/LC`, `A_AAC/MPEG2/LC/SBR`, `A_AAC/MPEG2/LC`, and `A_AAC/MPEG4/LC/SBR`. This should open up more test files on my end.

darksv replied:
Sure, I will add these.

pdeljanov commented on symphonia-format-mkv/src/lib.rs:
```diff
@@ -0,0 +1,329 @@
+                buffer.push_back((track, reader.read_boxed_slice_exact(frame_size)?));
+            }
+        }
+        _ => unreachable!(),
```
This catch-all arm shouldn't be needed since the match covers all cases. If we keep it, then any additions to the `Lacing` will fall through to this case and cause a panic in practice.

darksv replied:
Right. I missed it when introducing an enum.

pdeljanov commented on symphonia-format-mkv/src/lib.rs:
```diff
@@ -0,0 +1,329 @@
+                codec_params.with_channel_layout(match audio.channels {
+                    1 => Layout::Mono,
+                    2 => Layout::Stereo,
+                    _ => unimplemented!(),
```
This `unimplemented!()` case is one I've run into a lot. There's a few options here. You could just not provide the channel layout. Or, if there are no `Layout`s that match what's defined for >2 channels, you can use `CodecParameters::with_channels` and provide a `Channels` bitmask.

pdeljanov commented on symphonia-format-mkv/src/lib.rs:
```diff
@@ -0,0 +1,329 @@
+                        end: header.pos + header.element_len,
+                    });
+                }
+                ElementType::Timestamp => {
+                    assert!(self.current_cluster.is_some());
+                    if let Some(cluster) = &mut self.current_cluster {
+                        cluster.timestamp = Some(self.iter.read_u64().unwrap());
+                    }
+                }
+                ElementType::SimpleBlock => {
+                    assert!(self.current_cluster.is_some());
+                    let data = self.iter.read_boxed_slice().unwrap();
+                    extract_frames(&data, &mut self.frames).unwrap();
+                }
+                ElementType::BlockGroup => {
+                    assert!(self.current_cluster.is_some());
+                    let group = self.iter.read_element_data::<BlockGroupElement>().unwrap();
+                    extract_frames(&group.data, &mut self.frames).unwrap();
```
In general, `unwrap` needs to be avoided unless you can 100% be certain it won't panic in practice. I've run into a lot of panics when `extra_frames` is unwrapped.

darksv replied:
Sure, I'm going to improve error handling later

pdeljanov commented:
Hey @darksv,

The progress is looking good! Sadly, I haven't had much luck in getting anything playing on my end, but I think it's close. I've noted in the review comments where it's getting stuck on my end.

There are also some bugs that I need to fix as well. Particularly, `FlacDecoder` doesn't parse extra data and instead expects some `CodecParameters` fields to be set. This is something I will look that will hopefully get more files playing.

darksv commented:
I added vorbis and remaining AAC codecs. Also fixed two off-by-one errors in `extract_frames`

pdeljanov commented:
Having very good results with the latest set of changes! Many of the MKVs I have with Vorbis or AAC audio tracks are working pretty flawlessly, even ones with multiple tracks. Lot's of debug logs though :sweat_smile: .

I'll push a commit into the main branch that fixes the issue with FLAC tracks in the next day or so.

pdeljanov commented:
I've fixed the issue with `FlacDecoder`, however I need some support in `MkvReader` to come to the optimal solution.

For the optimal solution, the extra data for FLAC should just contain the Stream Info block. However, the codec private data provided by MKV contains many of the FLAC metadata blocks which are useless for the decoder and may bloat the extra data substantially if there's stuff like album art embedded into the file.

Could you make it so that only the Stream Info block is provided in the extra data?

The following code should extract the Stream Info block from the MKV codec private data which you can then add to the CodecParameters.

```rust
use symphonia_utils_xiph::flac::metadata::{MetadataBlockHeader, MetadataBlockType};

fn get_stream_info_from_codec_private(codec_private: &[u8]) -> Result<Box<[u8]>> {
    let mut reader = BufReader::new(codec_private);

    let marker = reader.read_quad_bytes()?;

    if marker != *b"fLaC" {
        return unsupported_error("mkv (flac): missing flac stream marker");
    }

    let header = MetadataBlockHeader::read(&mut reader)?;

    loop {
        match header.block_type {
            MetadataBlockType::StreamInfo => {
                break Ok(reader.read_boxed_slice_exact(header.block_len as usize)?);
            }
            _ => reader.ignore_bytes(u64::from(header.block_len))?,
        }
    }
}
```

darksv commented:
I've just pushed the commit with this change.

pdeljanov commented:
I pushed the changes required for FLAC to work. It should be good the next time you rebase.

pdeljanov commented:
Hey @darksv,

Let me know when you think this may be ready for an initial merge. I've been a bit busy with life and other bug reports lately so I haven't been able to keep up with these changes.

When you think it may be good to go I can do a more thorough review and suggest changes to get it ready for merging.

darksv commented:
@pdeljanov I think it might be ready for a more detailed review. Right now it still has some rough edges, but it seems to work with the files I tested.

CRC validation is not implemented yet and I need to improve handling of broken files. I believe it could be done later, after merging this PR.

pdeljanov commented:
Hey @darksv,

First off, thank you for your hard work on this. I'd love to publish this in the upcoming version v0.5.0 release.

Aside from the review, before I can merge this there are a few issues that need to be resolved:

1. With larger files, it takes a very long time before playback starts. This doesn't occur when I provide the file via. standard input. I'm guessing the reader is scanning all the clusters if the input is seekable? Can we avoid that or make it more optimized?
2. The code must compile without any warnings. You can use `#[allow(dead_code)]` on structs with fields you are planning to use later. Prefixing unused variables with `_` is also acceptable.
3. There's a lot of "Ignored Element" warnings (10000+, it exceeded my terminal's scrollback). Particularly CueRelativePosition, CueDuration, TagLanguage, and TagDefault. These should be debug level logs to avoid spamming the end-user's logs. Just in general, warnings should only be used for things that have a good chance of impacting the user.
4. Not really an issue, but please fix any `clippy` lints.

While you look into those, I'll start working on a thorough review.

Thanks again!

darksv commented:
I've fixed all the warnings, including clippy lints. Also changed warnings about ignored elements to debug messages.

> With larger files, it takes a very long time before playback starts. This doesn't occur when I provide the file via. standard input. I'm guessing the reader is scanning all the clusters if the input is seekable? Can we avoid that or make it more optimized?

Thanks for pointing that out. I will work on this soon.

pdeljanov reviewed (changes requested):
I'm probably going to have to do a second round, there's a lot of code here! :sweat_smile:

Just some general comments:

1) In the current definition of the API, any error returned by `FormatReader`is basically fatal to demuxing (aside from `ResetRequired`). Please make sure that if you return an error that it's not possible to continue by skipping/ignoring the damaged element, cluster, or track. Of course, there is balance to be had here. We shouldn't complicate the code to try and handle completely messed up files, but if it's possible to be more tolerant of damaged files we should try. I made comments in a few locations where I think this could be done.

2) Since Symphonia is a library, we should not do anything that can panic (e.g., `assert`, `unwrap`, etc.). Anything fatal should return an error. I've added comments where I think this may be a problem. Using `debug_assert` should be fine, but still should be reserved for truly exceptional circumstances.

Thanks!

darksv commented on symphonia-format-mkv/src/ebml.rs:
```diff
@@ -0,0 +1,443 @@
+impl ElementHeader {
+    /// Returns an iterator over child elements of the current element.
+    pub(crate) fn children<R: ReadBytes>(&self, reader: R) -> ElementIterator<R> {
+        assert_eq!(reader.pos(), self.data_pos, "unexpected position");
```
No. This assertion should always pass unless there is a bug in the code. `children` is called always after reading the EBML header so stream position is valid. I should probably add a comment about it.

darksv commented on symphonia-format-mkv/src/segment.rs:
```diff
@@ -0,0 +1,622 @@
+            duration,
+            title: title.map(|it| it.into_boxed_str()),
+            muxing_app: muxing_app.ok_or(Error::Unsupported("mkv: missing muxing app"))?.into_boxed_str(),
+            writing_app: writing_app.ok_or(Error::Unsupported("mkv: missing muxing app"))?.into_boxed_str(),
```
Done.

darksv commented on symphonia-format-mkv/src/demuxer.rs:
```diff
@@ -0,0 +1,511 @@
+    let mut setup_header = None;
+
+    for packet in packets {
+        match packet[0] {
```
Added check

darksv commented on symphonia-format-mkv/src/demuxer.rs:
```diff
@@ -0,0 +1,511 @@
+    ].concat().into_boxed_slice())
+}
+
+fn get_stream_info_from_codec_private(codec_private: &[u8]) -> Result<Box<[u8]>> {
```
done

darksv commented on symphonia-format-mkv/src/demuxer.rs:
```diff
@@ -0,0 +1,511 @@
+
+    let marker = reader.read_quad_bytes()?;
+    if marker != *b"fLaC" {
+        return unsupported_error("mkv (flac): missing flac stream marker");
```
changed to `DecodeError`

darksv commented on symphonia-format-mkv/src/ebml.rs:
```diff
@@ -0,0 +1,443 @@
+/// Reads a single EBML element ID (as in RFC8794) from the stream
+/// and returns it or an error.
+pub(crate) fn read_tag<R: ReadBytes>(reader: R) -> Result<u32> {
+    Ok(read_vint::<R, false>(reader)?.0 as u32)
```
added validation

darksv commented on symphonia-format-mkv/src/ebml.rs:
```diff
@@ -0,0 +1,443 @@
+            self.reader.ignore_bytes(self.next_pos - pos)?;
+        }
+
+        assert_eq!(self.next_pos, self.reader.pos(), "invalid position");
```
It shouldn't

darksv commented on symphonia-format-mkv/src/ebml.rs:
```diff
@@ -0,0 +1,443 @@
+                header
+            }
+        };
+        assert_eq!(header.etype, E::ID, "reading invalid element");
```
This is to make sure that this function is always called after reading an EBML header and checking it's type. It shouldn't fail.

darksv replied:
I rewrote this function so it doesn't read the EBML header implicitly and improved messages inside the assertions.

pdeljanov commented:
> 2. I've run into some MKVs that don't play at all. I get the following error instead:
>
>    `ERROR symphonia_play > file not supported. reason? end of stream`
>
>    It's kind of weird because the file belongs to a series and atleast some of the other episodes play fine. The file is pretty large so not sure if I can share it with you. Let me know if there's any information I can provide.

Oh, funny, after your changes, I now get `ERROR symphonia_play > file not supported. reason? malformed stream: mkv: too long element ID` for this file.

darksv commented:
Sorry, last week I wasn't able to work on this. I'll try to get back to fixing remaining issues today.

> Oh, funny, after your changes, I now get `ERROR symphonia_play > file not supported. reason? malformed stream: mkv: too long element ID` for this file.

Interesting... I'll look into it. This probably should be fixed by skipping data in the stream until a valid tag is found.

darksv commented:
@pdeljanov I've pushed some changes with handling of clusters with unknown sizes. Let me know if this fixes your issue.

pdeljanov commented:
Hi @darksv,

No problem, take your time!

I get a panic instead of an error now. Please see the backtrace below:

```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `562413392`,
 right: `562426428`: invalid position', symphonia-format-mkv/src/ebml.rs:286:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/02072b482a8b5357f7fb5e5637444ae30e423c40/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /rustc/02072b482a8b5357f7fb5e5637444ae30e423c40/library/core/src/panicking.rs:107:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/02072b482a8b5357f7fb5e5637444ae30e423c40/library/core/src/panicking.rs:145:5
   4: symphonia_format_mkv::ebml::ElementIterator<R>::read_header_no_consume
             at ./symphonia-format-mkv/src/ebml.rs:286:9
   5: symphonia_format_mkv::ebml::ElementIterator<R>::read_header
             at ./symphonia-format-mkv/src/ebml.rs:249:22
   6: <symphonia_format_mkv::segment::ClusterElement as symphonia_format_mkv::ebml::Element>::read
             at ./symphonia-format-mkv/src/segment.rs:490:34
   7: symphonia_format_mkv::ebml::ElementIterator<R>::read_element_data
             at ./symphonia-format-mkv/src/ebml.rs:310:23
   8: <symphonia_format_mkv::demuxer::MkvReader as symphonia_core::formats::FormatReader>::try_new
             at ./symphonia-format-mkv/src/demuxer.rs:333:35
   9: <symphonia_format_mkv::demuxer::MkvReader as symphonia_core::probe::QueryDescriptor>::query::{{closure}}
             at ./symphonia-core/src/probe.rs:351:65
  10: core::ops::function::FnOnce::call_once
             at /rustc/02072b482a8b5357f7fb5e5637444ae30e423c40/library/core/src/ops/function.rs:227:5
  11: symphonia_core::probe::Probe::format
             at ./symphonia-core/src/probe.rs:317:34
  12: symphonia_play::main
             at ./symphonia-play/src/main.rs:129:11
  13: core::ops::function::FnOnce::call_once
             at /rustc/02072b482a8b5357f7fb5e5637444ae30e423c40/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

darksv commented:
@pdeljanov Please check whether recent changes fix this.

pdeljanov commented:
Nice work, @darksv, the file plays now!

darksv commented on symphonia-format-mkv/src/demuxer.rs:
```diff
@@ -0,0 +1,511 @@
+                    }
+                    None => {
+                        self.iter.ignore_data()?;
+                        return decode_error("mkv: timestamp element outside of a cluster");
```
I replaced `DecodeError`s with just warnings.

darksv commented on symphonia-format-mkv/src/demuxer.rs:
```diff
@@ -0,0 +1,511 @@
+                let cluster = match self.current_cluster.as_ref() {
+                    Some(cluster) => cluster,
+                    None => {
+                        self.iter.ignore_data()?;
```
also fixed

darksv commented on symphonia-format-mkv/src/demuxer.rs:
```diff
@@ -0,0 +1,511 @@
+                let cluster = match self.current_cluster.as_ref() {
+                    Some(cluster) => cluster,
+                    None => {
+                        self.iter.ignore_data()?;
```
also fixed

darksv commented on symphonia-format-mkv/src/segment.rs:
```diff
@@ -0,0 +1,622 @@
+                    bit_depth = Some(it.read_u64()?);
+                }
+                other => {
+                    log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/segment.rs:
```diff
@@ -0,0 +1,622 @@
+                    seeks.push(it.read_element_data()?);
+                }
+                other => {
+                    log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/segment.rs:
```diff
@@ -0,0 +1,622 @@
+                    seek_position = Some(it.read_u64()?);
+                }
+                other => {
+                    log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/segment.rs:
```diff
@@ -0,0 +1,622 @@
+                    doc_type_read_version = Some(it.read_u64()?);
+                }
+                other => {
+                    log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/segment.rs:
```diff
@@ -0,0 +1,622 @@
+                    writing_app = Some(it.read_string()?);
+                }
+                other => {
+                    log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/segment.rs:
```diff
@@ -0,0 +1,622 @@
+                    pos = Some(it.read_element_data()?);
+                }
+                other => {
+                    log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/segment.rs:
```diff
@@ -0,0 +1,622 @@
+                    pos = Some(it.read_u64()?);
+                }
+                other => {
+                    log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/segment.rs:
```diff
@@ -0,0 +1,622 @@
+                    block_duration = Some(it.read_u64()?);
+                }
+                other => {
+                    log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/segment.rs:
```diff
@@ -0,0 +1,622 @@
+                    tags.push(it.read_element_data::<TagElement>()?);
+                }
+                other => {
+                    log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/segment.rs:
```diff
@@ -0,0 +1,622 @@
+                    simple_tags.push(it.read_element_data::<SimpleTagElement>()?);
+                }
+                other => {
+                    log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/segment.rs:
```diff
@@ -0,0 +1,622 @@
+                    value = Some(it.read_data()?);
+                }
+                other => {
+                    log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/demuxer.rs:
```diff
@@ -0,0 +1,511 @@
+                self.metadata.push(tags.to_metadata());
+            }
+            other => {
+                log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/demuxer.rs:
```diff
@@ -0,0 +1,511 @@
+                    clusters.push(it.read_element_data::<ClusterElement>()?);
+                }
+                other => {
+                    log::warn!("ignored element {:?}", other);
```
done

darksv commented on symphonia-format-mkv/src/demuxer.rs:
```diff
@@ -0,0 +1,511 @@
+                    }
+                };
+
+                let timestamp = cluster.timestamp.ok_or(Error::DecodeError("mkv: missing cluster timestamp"))?;
```
changed to `log::warn` and skipping the cluster

darksv commented on symphonia-format-mkv/src/demuxer.rs:
```diff
@@ -0,0 +1,511 @@
+                }
+            }
+
+            let track_id = u32::try_from(track.number).unwrap();
```
fixed

darksv commented on symphonia-format-mkv/src/ebml.rs:
```diff
@@ -0,0 +1,443 @@
+            Some((ty, _)) => {
+                assert_eq!(header.data_pos, self.reader.pos(), "invalid stream position");
+                if let (Some(cur), Some(end)) = (self.current, self.end) {
+                    assert!(cur.pos + cur.len <= end, "invalid stream position");
```
replaced this assertion with a `DecodeError`

darksv commented on symphonia-format-mkv/src/ebml.rs:
```diff
@@ -0,0 +1,443 @@
+                // TODO: ignore crc for now
+                continue;
+            }
+            assert_eq!(header.etype, E::ID);
```
replaced with just a `log::warn`

darksv commented on symphonia-format-mkv/src/ebml.rs:
```diff
@@ -0,0 +1,443 @@
+    pub(crate) fn try_read_data(&mut self, header: ElementHeader) -> Result<Option<ElementData>> {
+        Ok(match ELEMENTS.get(&header.tag) {
+            Some((ty, _)) => {
+                assert_eq!(header.data_pos, self.reader.pos(), "invalid stream position");
```
added comment

pdeljanov commented:
Hi @darksv,

I think for the most part all the comments have been addressed. There are two issues though:

1. It's very slow to open a file (for me it's taking ~30 seconds for a moderately sized video).
2. On end-of-stream the demuxer returns a `DecodeError` rather than an `IoError`.

Are those going to be easy to fix?

I'd like to publish v0.5 early next week. However, if you think it would take longer than that to fix those issues, then we can target v0.5.1 for MKV support instead. I feel the first issue may be a show-stopper for many users so it probably wouldn't make sense to include it in v0.5 as-is.

darksv commented:
@pdeljanov Both issues should be fixed now, but I'll have to test it a little bit more.

pdeljanov commented:
Hi @darksv, it opens real quick now!

However, the method you're using to try to generate an end-of-stream error causes an OS error to occur for me (Linux 5.16.0):
```
ERROR symphonia_play > error: Invalid argument (os error 22)
```
I've never seen that kind of error before, but I believe it's because `ignore_bytes` will issue a seek if the amount of bytes exceeds the size of `MediaSourceStream`'s internal buffer rather than read through the file sequentially. The OS probably doesn't like the seek forward of 2^64 bytes.

This is easily fixed though, your `assert_end_of_stream` function can either be replaced with, or simply call, `symphonia_core::errors::end_of_stream_error()` to generate the correct error.

darksv commented:
@pdeljanov Thanks! I don't know how I missed this function... 😄

pdeljanov commented:
No worries!

I think what you've implemented right now is very good and usable. There will probably be further developments, but I think we can address those as smaller incremental PRs.

For the 0.5 release we can classify the `MkvReader` as "Good." That way we can get more people testing it while also setting reasonable expectations for a new and large component.

There are just some last remaining tasks that need to be done:

1. Rebase to the latest `master`. `FormatReader` was make `Sync` recently, so that may require some small changes on your end.
2. Update `Cargo.toml` for `symphonia-format-mkv`. You should be the author, and the keywords and version fields need to be updated. You can use the version number `0.4.0` so when I mass update the version numbers to `0.5.0` it gets picked up.
3. Run `cargo clippy` one last time
4. Run `rustfmt`. The configuration we use requires the nightly toolchain so you'd need to install that then run `cargo +nightly fmt`. If this is too burdensome, then I can run it after merging.

Once you address those items and make this PR not a draft, I'll merge it.

pdeljanov commented:
Thank you @darksv! I, and I'm sure many other people, appreciate the work you've put into implementing this.
